src/lib/Hydra/Controller/User.pm

```diff
@@ -143,6 +157,38 @@ sub doEmailLogin {
 }
 
 
+sub oidc_login :Path('/oidc-login') Args(0) {
+    my ($self, $c) = @_;
+
+    error($c, "Logging in via OIDC is not enabled.", 404) unless $c->config->{enable_oidc_login};
+    my $oidc_client = get_oidc_client($c);
+
+    error($c, q{OIDC state mismatch. Is this a CSRF attack?}) unless $c->session->{oidc_state} && $c->session->{oidc_state} eq $c->request->param("state");
+
+    my $code = $c->request->param("code");
+    my $token_response = $oidc_client->get_access_token(
+        code => $code,
+        redirect_uri => $c->uri_for('/oidc-login'),
+    ) or error($c, $oidc_client->errstr);
+
+    $c->session->{access_token} = $token_response->access_token;
+    $c->session->{expires_at} = time() + $token_response->expires_in;
+    $c->session->{refresh_token} = $token_response->refresh_token;
+
```

Comment:
We really ought to check the `id_token` and its signature here and make sure that the `id_token` is bound to the audience (client_id)..  And then save the `sub` and compare it to the sub received in userinfo.

Otherwise we are vulnerable against token substitution attacks as the access token is not audience-bound.

> NOTE: Due to the possibility of token substitution attacks, the UserInfo Response is not guaranteed to be about the End-User identified by the sub (subject) element of the ID Token. The sub Claim in the UserInfo Response MUST be verified to exactly match the sub Claim in the ID Token; if they do not match, the UserInfo Response values MUST NOT be used. 

https://openid.net/specs/openid-connect-basic-1_0.html#UserInfo

We're not reusing the `access_token` long-term but still. we should use the `id_token` as the source of truth about the user identity. Not the `userinfo` response